common/deliverclient/block_verification.go

```diff
@@ -0,0 +1,350 @@
+	VerifyBlockAttestation(block *common.Block) error
+
+	// UpdateConfig sets the config by which blocks are verified. It is assumed that this config block had already been
+	// verified using the VerifyBlock method immediately prior to calling this method.
+	UpdateConfig(configBlock *common.Block) error
+
+	// UpdateBlockHeader saves the last block header that was verified and handled successfully.
+	// This must be called after VerifyBlock and VerifyBlockAttestation and successfully handling the block.
+	UpdateBlockHeader(block *common.Block)
+
+	// Clone makes a copy from the current verifier, a copy that can keep being updated independently.
+	Clone() CloneableUpdatableBlockVerifier
+}
+
+// BlockVerificationAssistant verifies the integrity and signatures of a block stream, while keeping a copy of the
+// latest configuration.
+//
+// Every time a config block arrives, it must first be verified using VerifyBlock and then
+// used as an argument to the UpdateConfig method.
+// The block stream could be composed of either:
+// - full blocks, which are verified using the VerifyBlock method, or
+// - block attestations (a header+metadata, with nil data) which are verified using the VerifyBlockAttestation method.
+// In both cases, config blocks must arrive in full.
+type BlockVerificationAssistant struct {
+	channelID string
+
+	// Creates the sigVerifierFunc whenever the configuration is set or updated.
+	verifierAssembler *BlockVerifierAssembler
+	// Verifies block signature(s). Recreated whenever the configuration is set or updated.
+	sigVerifierFunc protoutil.BlockVerifierFunc
+	// The current config block header.
+	// It may be nil if the BlockVerificationAssistant is created from common.Config and not a config block.
+	// After 'UpdateConfig(*common.Block) error' this field is always set.
+	configBlockHeader *common.BlockHeader
+	// The last block header may include the number only, in case we start from common.Config.
+	lastBlockHeader *common.BlockHeader
+	// The last block header hash is given when we start from common.Config, and is computed otherwise.
+	lastBlockHeaderHash []byte
+
+	logger *flogging.FabricLogger
+}
+
+// NewBlockVerificationAssistant creates a new BlockVerificationAssistant from a config block.
+// This is used in the orderer, where we always have access to the last config block.
+func NewBlockVerificationAssistant(configBlock *common.Block, lastBlock *common.Block, cryptoProvider bccsp.BCCSP, lg *flogging.FabricLogger) (*BlockVerificationAssistant, error) {
+	if configBlock == nil {
+		return nil, errors.Errorf("config block is nil")
+	}
+	if configBlock.Header == nil {
+		return nil, errors.Errorf("config block header is nil")
+	}
+	configIndex, err := protoutil.GetLastConfigIndexFromBlock(configBlock)
+	if err != nil {
+		return nil, errors.WithMessage(err, "error getting config index from config block")
+	}
+	if configIndex != configBlock.Header.Number {
```

Comment:
correct, I added a check
```
	if !protoutil.IsConfigBlock(configBlock) {
		return nil, errors.New("config block parameter does not carry a config block")
	}
```

the reason I am checking the index is because I am verifying that the last block refers to the same config block we are starting with.